### ISSUE TITLE
Different pagination behaviour using Array connection wrapper vs ActiveRecord Associations

### ISSUE THREAD

**Issue by joeytepp** [rmosolgo/graphql-ruby#4765]
**Describe the bug**

When returning a paginated connection field, the behaviour is different when an array of ActiveRecord objects is returned vs an ActiveRecord association. This is most noticeable when both `before` and `after` are supplied, and the corresponding indices to the cursors are 1 apart (ie the index corresponding to the `after` cursor is 1 smaller than the index corresponding to the `before` cursor). When this happens we see a single edge in the payload where we would expect an empty array.

**Versions**

`graphql` version: 2.1.6
`rails` (or other framework): 7.1.1

**GraphQL schema**

Include relevant types and fields (in Ruby is best, in GraphQL IDL is ok). Any custom extensions, etc?

```ruby
class Product < GraphQL::Schema::Object
  field :id, ID, hash_key: :id
  # …
end

class QueryType < GraphQL::Schema::Object
  field :products_using_array, Product.connection_type, null: false

  def products_using_array
    ::Product.all.to_a
  end

  field :products_using_association, Product.connection_type, null: false

  def products_using_association
    ::Product.all
  end
end

class ApplicationSchema < GraphQL::Schema
  query QueryType
  # …
end
```

**GraphQL query**

Example GraphQL query and response (if query execution is involved)

```graphql
query {
  productsUsingArray(
    after: "MQ", # Index 1
    before: "Mg"  # Index 2
  ) {
    products {
      edges {
        cursor
        node {
           ...
        }
     }
  }
}
```
The query above will return the following payload (with page info included)
```json
{
  "data": {
    "features": {
      "pageInfo": {
        "startCursor": "Mg",
        "endCursor": "Mg",
        "hasNextPage": false,
        "hasPreviousPage": true
      },
      "edges": [
        {
          "cursor": "Mg",
          "node": {
              ...
          }
        }
      ]
    }
  }
}
```

Notice how the edge with cursor `Mg` still appears in the payload, despite the fact that we've specified it as our `before` value (which is an exclusive boundary).

When we repeat the same query but using the ActiveRecord connection wrapper instead of the Array-based one we get a different result

```json
{
  "data": {
    "features": {
      "pageInfo": {
        "startCursor": null,
        "endCursor": null,
        "hasNextPage": true,
        "hasPreviousPage": true
      },
      "edges": []
    }
  }
}
```

This is the result I would expect from such a query, as the exclusive boundary should not have any records in its range. It seems to me like the `before` value is being ignored in this case, will take a closer look and see if I can put up a fix tomorrow!

**Comment by rmosolgo** [comment]
Hey, thanks for the detailed report! I agree they should both return an empty list of edges.

Maybe the range slice here should use `...` (excluding the last item) instead of `..` (including the last item): 

https://github.com/rmosolgo/graphql-ruby/blob/d93e2fbab1747d46ec1089bab56e72875fd02c89/lib/graphql/pagination/array_connection.rb#L39

**Comment by joeytepp** [comment]
Yes, I initially looked at that as a solution, the `has_next_page` and `has_prev_page` values also need to be updated too though which becomes tricky for this particular case.

**Comment by rmosolgo** [comment]
I've got a fix in the works at #4780 -- I think it was really two bugs: 

- The final-index bug in the array slicing 
- `hasNextPage` didn't follow the Relay spec perfectly -- it _should_ be `false` in that case
